### PR TITLE
Fix policy DB corruption on ignition cycle

### LINKED ISSUE #2467: Policy table corupted after first ign cycle
### Bug Report
Policy table corupted after first ign cycle

##### Preconditions:
1. First SDL life cycle .
2. HMI, SDL are started.
3. Device is paired via BT.

##### Reproduction Steps
1. Register app SyncProxyTester, 584421907.
2. Activate app.
3. Consent device.
4. Perform PTU.
5. Make Ign_off-on -> HU is restarted.
6. Register app.

##### Expected Behavior
Permissions should not be lost.

##### Observed Behavior
SDL disallowes all RPCs.

##### OS & Version Information
* OS/Version: 
* SDL Core Version: 
* Testing Against:

### PR DESCRIPTION
**Pull request by LitvinenkoIra**
Fixes #2467

This PR is **[ready]** for review.

### Risk
This PR makes **[no]** API changes.

### Summary
Due to data races on ignition on because of SDL starts getting messages in the middle of policy initialization and on ignition off because of backup thread wasn't stopped after OnSDLPersistenceComplete policy database was occasionally corrupting.
In order to prevent that handling of SUSPEND/AWAKE signals has been added as well as policy loading state check in ignition on.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by LitvinenkoIra** [comment]
note: commit 6d3f840eb03f1585c4e77f2e15a909378d0cadb1 contains not related changes, but I need it for successful build with external policy

File: src/components/policy/policy_external/src/cache_manager.cc [hunk 1]
```diff
@@ -775,7 +776,15 @@ bool CacheManager::AppHasHMIType(const std::string& application_id,
 
 void CacheManager::Backup() {
   sync_primitives::AutoLock lock(backuper_locker_);
-  DCHECK(backuper_);
+  if (NULL == backup_thread_) {
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/src/cache_manager.cc]: @LitvinenkoIra is it regular situation that there is no backup_thread? May be `DCHECK_OR_RETURN_VOID(backup_thread_)`

File: src/components/policy/policy_external/include/policy/cache_manager.h [hunk 2]
```diff
@@ -955,6 +958,7 @@ class CacheManager : public CacheManagerInterface {
   sync_primitives::Lock backuper_locker_;
   BackgroundBackuper* backuper_;
   const PolicySettings* settings_;
+  volatile bool is_policy_table_loaded_;
```
Comment (ZhdanovP) [inline src/components/policy/policy_external/include/policy/cache_manager.h]: @LitvinenkoIra Do we need `volatile` here , does this variable change from other thread?

File: src/components/policy/policy_external/src/cache_manager.cc [hunk 3]
```diff
@@ -2745,11 +2765,24 @@ void CacheManager::MergeCFM(const policy_table::PolicyTable& new_pt,
...
+  sync_primitives::AutoLock lock(backuper_locker_);
   backuper_ = new BackgroundBackuper(this);
   backup_thread_ = threads::CreateThread("Backup thread", backuper_);
   backup_thread_->start();
 }
 
+void CacheManager::StopBackupTread() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  sync_primitives::AutoLock lock(backuper_locker_);
+  if (NULL == backup_thread_) {
```
Comment (ZhdanovP) [inline src/components/policy/policy_external/src/cache_manager.cc]: @LitvinenkoIra let's use `nullptr`

File: src/components/policy/policy_external/include/policy/policy_types.h [hunk 4]
```diff
@@ -531,6 +531,16 @@ enum ConsentPriorityType { kUserConsentPrio, kExternalConsentPrio };
  */
 enum ConsentProcessingPolicy { kTimestampBased, kExternalConsentBased };
 
+/**
+ * @brief The SystemState enum defines possible system state what policy has to
+ * be aware of
+ */
+enum class SystemState : uint32_t {
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/include/policy/policy_types.h]: @LitvinenkoIra would be nice to describe what events moves SDL to what state,

File: src/components/policy/policy_external/src/cache_manager.cc (hunk 1 above)
Comment (LuxoftAKutsan) [reply src/components/policy/policy_external/src/cache_manager.cc]: @LitvinenkoIra any updates?

File: src/components/policy/policy_external/src/cache_manager.cc (hunk 3 above)
Comment (LuxoftAKutsan) [reply src/components/policy/policy_external/src/cache_manager.cc]: @LitvinenkoIra check if this condition should be checked with DCHECK

File: src/components/policy/policy_external/src/cache_manager.cc [hunk 5]
```diff
@@ -2781,6 +2814,20 @@ void CacheManager::OnDeviceSwitching(const std::string& device_id_from,
...
+void CacheManager::OnSystemStateChanged(SystemState state) {
+  switch (state) {
+    case SystemState::kStateSuspended:
+      StopBackupTread();
+      break;
+    case SystemState::kStateAwaken:
+      InitBackupThread();
+      break;
+    default:
+      LOG4CXX_ERROR(logger_, "Unknown system state.");
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/src/cache_manager.cc]: @LitvinenkoIra what about default ready state?

File: src/components/policy/policy_regular/src/cache_manager.cc [hunk 6]
```diff
@@ -1760,6 +1768,40 @@ void CacheManager::OnDeviceSwitching(const std::string& device_id_from,
...
+  sync_primitives::AutoLock lock(backuper_locker_);
+  backuper_ = new BackgroundBackuper(this);
+  backup_thread_ = threads::CreateThread("Backup thread", backuper_);
+  backup_thread_->start();
+}
+
+void CacheManager::StopBackupTread() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  sync_primitives::AutoLock lock(backuper_locker_);
+  if (NULL == backup_thread_) {
```
Comment (AByzhynar) [inline src/components/policy/policy_regular/src/cache_manager.cc]: @LitvinenkoIra Use `nullptr`

**Comment by JackLivio** [comment]
@LitvinenkoIra @AByzhynar This Pr looks quite stale, has a lot of merge conflicts, and unresolved comments. Let me know if I can close this PR.

**Comment by jacobkeeler** [comment]
Issue was resolved separately by #2609

**Pull request closed without merging**